language/move-compiler/src/typing/core.rs

```diff
@@ -1335,13 +1352,20 @@ fn make_tparams(
     case: TVarCase,
     tparam_constraints: Vec<(Loc, AbilitySet)>,
 ) -> Vec<Type> {
+    let arity = tparam_constraints.len();
     tparam_constraints
         .into_iter()
-        .map(|(vloc, constraint)| {
+        .enumerate()
+        .map(|(i, (vloc, constraint))| {
             let tvar = make_tvar(context, vloc);
             context.add_ability_set_constraint(loc, None::<String>, tvar.clone(), constraint);
             match &case {
-                TVarCase::Single(msg) => context.add_single_type_constraint(loc, msg, tvar.clone()),
+                TVarCase::SingleButLast(_) if i + 1 == arity => {
+                    // Last arg (return type of a function): do not add any constraint
+                }
+                TVarCase::SingleButLast(msg) | TVarCase::Single(msg) => {
```

Comment:
What does `SingleButLast` mean when it's not the return type? Would it be possible to call this enum variant `TVarCase::ReturnType` and raise an error when it's not the return type?

Reply:
I renamed this to `Function` for clarity. The meaning of it is as the code says: for any argument which is not a return value, it behaves like Single, for the return argument it is unconstrained.